Dev testing

itsHaseebSaeed opened the pull request:
# Privacy-Enhanced Trainer Commitments (Swarm V2 Upgrade)

### Highlights

- Introduced a **Swarm V2 stack** that layers privacy-preserving trainer commitments on top of the existing training lifecycle.
- Aggregators can now opt into a **privacy mode**: they lodge trainer commitments during training, reserve bond per commitment, and reveal identities later.
- Introduced **bond-backed enforcement** with configurable penalties and finder rewards to discourage missed reveals.

### Contract Changes


- `forge/src/swarm/registry/RoundTrainerRegistryV2.sol`
  - Replaces the simple trainer registry with namespaced storage for both trainer metadata and privacy commitments.
  - Implements `_commitTrainerPrivacy`, `_revealTrainerPrivacy`, `_slashTrainerCommitment`, bond accounting, and configuration setters.
  - Exposes read helpers for trainer info, bond state, pending commitment counts, and privacy configuration.

  - `forge/src/swarm/SwarmV2.sol`
  - Extends Swarm by inheriting the new `RoundTrainerRegistryV2` with commitment lifecycle helpers.
  - Adds aggregator-only APIs: `registerRoundContributionPrivacy`, `revealTrainerCommitment`, `slashTrainerCommitment`, `configureTrainerPrivacy`, bond deposit/withdraw, and a privacy-mode toggle.
  - Guards evaluator assignments and reward calculations to include pending commitments so tasks scale with unrevealed trainers.

- `forge/src/swarm/SwarmV2Factory.sol`
  - Mirrors the V1 factory but wires in the Swarm V2 logic address and reuses existing selector/calculator/access/compensation factories.


### Scripts & Demo Assets

- Demo scaffolding (`forge/test/training/DemoParams.sol`, `forge/test/training/FullCyclePrivacy.t.sol`) now orchestrates a privacy-enabled round: commitments, evaluator registration, reveals, deliberate missed reveal, slashing, and reward distribution.

### Tests

- `forge/test/swarm/SwarmV2.t.sol`
  - Validates access control modifiers, full training lifecycle, privacy-mode interactions (commit/reveal/slash), and reward flows with pending commitments.
- `forge/test/swarm/registry/RoundTrainerRegistryV2.t.sol`
  - Unit tests commitment storage, bond accounting, model hash updates, and edge cases around duplicate/late operations.
- `forge/test/training/FullCyclePrivacy.t.sol`
  - Scenario test exercising the entire privacy demo, including slashing and reward distribution.
- Existing V1 tests remain untouched but still relevant for regression comparison.

### Testing Guidance

- Run the targeted suite to validate the upgrade:
  - `forge test --match-contract SwarmV2Test`
  - `forge test --match-contract RoundTrainerRegistryTest`
  - `forge test --match-contract FullCyclePrivacy`

itsHaseebSaeed committed:
feat: added tests scripts

itsHaseebSaeed committed:
feat: implement privacy mode for trainer contributions

itsHaseebSaeed committed:
chore: remove unused images and scripts, update .gitignore to exclude experimental results

itsHaseebSaeed committed:
chore: remove deprecated demo scripts and privacy mode functionality from SwarmV1

itsHaseebSaeed committed:
feat: add RoundTrainerRegistry contract for managing trainers per round

itsHaseebSaeed committed:
feat: enhance SwarmV2 and RoundTrainerRegistry with privacy commitment tracking and aggregator bond management

itsHaseebSaeed committed:
refactor: remove unused privacy-related structures and error handling from RoundTrainerRegistry

itsHaseebSaeed committed:
refactor: clean up MockRoundTrainerRegistry by removing unused privacy-related functions

itsHaseebSaeed committed:
feat: update SwarmV2 and RoundTrainerRegistry to use RoundTrainerRegistryV2, enhancing privacy commitment tracking and aggregator bond management

Mick00 commented on forge/src/swarm/SwarmV2.sol:
```diff
@@ -0,0 +1,422 @@
+
+    function registerRoundContributionPrivacy(uint256 roundId, bytes32 commitment, bytes32 modelHash, uint64 revealDeadline) external onlyAggregator(msg.sender) {
+        if (!_trainerPrivacyEnabled) {
+            revert PrivacyModeDisabled();
```
This feature around _trainerPrivacyEnabled doesn't seem required and adds a bit of extra steps and code. If the objective is to wait for the `configureTrainerPrivacy` and  `setTrainerPrivacyMode` to be called before starting training, I believe you don't need this.

The flow should simply be:
1. (already implemented) The swarm starts in IDLE_PHASE. Training is not started as it reverts on the next if. 
2. call `configureTrainerPrivacy`
3. (already implemented) call `startTrainingRound`

itsHaseebSaeed replied:
Well it's used for the backward compatibilty. It allows anyone to select private or public mode

Mick00 commented on forge/src/swarm/SwarmV2.sol:
```diff
@@ -0,0 +1,422 @@
+        if (roundId != currentRound()) {
+            revert ForbiddenRound(roundId);
+        }
+        _commitTrainerPrivacy(roundId, commitment, modelHash, revealDeadline);
```
There's no input validation on the revealDeadline which could make it possible the aggregator puts the deadline in 10 years.

I believe it has disincentives for that, but please confirm and document if that's the case.

The aggregator bond amount would be reserved/locked for 10 years in this case, which disincentivise that behavior.

itsHaseebSaeed replied:
fixed. Added a max Deadline of 2 days.

Mick00 commented on forge/src/swarm/SwarmV2.sol:
```diff
@@ -0,0 +1,422 @@
+        _revealTrainerPrivacy(roundId, trainer, nonce);
+    }
+
+    function slashTrainerCommitment(uint256 roundId, bytes32 commitment) external returns (uint256 penalty, uint256 finderReward) {
```
I'm being a little picky here, but slashTrainerCommitment leads to thinking you slash the trainer, but it actually slash the aggregator who miss the deadline.

I would rename that function

itsHaseebSaeed replied:
fixed

Mick00 commented on forge/src/swarm/SwarmV2.sol:
```diff
@@ -0,0 +1,422 @@
+    function registerRoundContribution(uint256 roundId, bytes32 modelHash) external {
+        if (_trainerPrivacyEnabled) {
+            revert PrivacyModeEnabled();
+        }
```
OK, so the boolean switch is for backward compatibility, but I think it would be cleaner getting rid of this function and other functionalities which aren't required for your specific implementation

itsHaseebSaeed replied:
yes exactly, I thought since swarmV2 is updated version it should allow users to have public setup as well.

Mick00 reviewed (commented):
Thanks for this contribution, it is a really interesting feature!

If I am not mistaken, I think the only thing missing would be an end to end round test.

Then we can merge this PR and start working on the integration with the Python side of things

itsHaseebSaeed committed:
refactor: rename slashTrainerCommitment to slashAggregatorBond

Updated:
- SwarmV2.sol: Renamed public function
- SwarmV2.t.sol: Updated test calls
- FullCyclePrivacy.t.sol: Updated test call
